pkg/vsphere/simulator/virtual_machine.go

```diff
@@ -0,0 +1,269 @@
+	vm.Summary.Runtime = vm.Runtime
+
+	err = vm.configure(spec)
+	if err != nil {
+		return nil, err
+	}
+
+	return vm, nil
+}
+
+func (vm *VirtualMachine) configure(spec *types.VirtualMachineConfigSpec) types.BaseMethodFault {
+	err := vm.configureDevices(spec)
+	if err != nil {
+		return err
+	}
+
+	apply := []struct {
+		src string
+		dst *string
+	}{
+		{spec.Name, &vm.Name},
+		{spec.Name, &vm.Config.Name},
+		{spec.GuestId, &vm.Config.GuestId},
+		{spec.GuestId, &vm.Config.GuestFullName},
+		{spec.GuestId, &vm.Summary.Guest.GuestId},
+		{spec.GuestId, &vm.Summary.Guest.GuestFullName},
+		{spec.Uuid, &vm.Config.Uuid},
+		{spec.Version, &vm.Config.Version},
+		{spec.Files.VmPathName, &vm.Config.Files.VmPathName},
+		{spec.Files.SnapshotDirectory, &vm.Config.Files.SnapshotDirectory},
+		{spec.Files.LogDirectory, &vm.Config.Files.LogDirectory},
+	}
+
+	for _, f := range apply {
+		if f.src != "" {
+			*f.dst = f.src
+		}
+	}
+
+	if spec.MemoryMB != 0 {
+		vm.Config.Hardware.MemoryMB = int32(spec.MemoryMB)
+		vm.Summary.Config.MemorySizeMB = vm.Config.Hardware.MemoryMB
+	}
+
+	if spec.NumCPUs != 0 {
+		vm.Config.Hardware.NumCPU = spec.NumCPUs
+		vm.Summary.Config.NumCpu = vm.Config.Hardware.NumCPU
+	}
+
+	vm.Config.Modified = time.Now()
+
+	return nil
+}
+
+func (vm *VirtualMachine) create(spec *types.VirtualMachineConfigSpec) types.BaseMethodFault {
+	p, fault := parseDatastorePath(spec.Files.VmPathName)
+	if fault != nil {
+		return fault
+	}
+
+	host := Map.Get(*vm.Runtime.Host).(*HostSystem)
+	ds := Map.FindByName(p.Datastore, host.Datastore).(*Datastore)
+
+	// Assuming local datastore for now.  TODO: should be via datastore browser / file manager
+	dir := ds.Info.GetDatastoreInfo().Url
+
+	vmx := path.Join(dir, p.Path)
+
+	_, err := os.Stat(vmx)
+	if err == nil {
+		return &types.FileAlreadyExists{
+			FileFault: types.FileFault{
+				File: vmx,
+			},
+		}
+	}
+
+	_ = os.MkdirAll(path.Dir(vmx), 0700)
+
+	f, err := os.Create(vmx)
+	if err != nil {
+		return &types.FileFault{
+			File: vmx,
+		}
+	}
+
+	_ = f.Close()
+
+	return nil
+}
+
+func (vm *VirtualMachine) configureDevices(spec *types.VirtualMachineConfigSpec) types.BaseMethodFault {
+	devices := object.VirtualDeviceList(vm.Config.Hardware.Device)
+
+	for i, change := range spec.DeviceChange {
+		dspec := change.GetVirtualDeviceConfigSpec()
+		device := dspec.Device.GetVirtualDevice()
+		invalid := &types.InvalidDeviceSpec{DeviceIndex: int32(i)}
+
+		switch dspec.Operation {
+		case types.VirtualDeviceConfigSpecOperationAdd:
+			if devices.FindByKey(device.Key) != nil {
+				return invalid
+			}
+			devices = append(devices, device)
+		}
+	}
+
+	vm.Config.Hardware.Device = []types.BaseVirtualDevice(devices)
+
+	return nil
+}
+
+type powerVMTask struct {
+	*VirtualMachine
+
+	state types.VirtualMachinePowerState
+}
+
+func (c *powerVMTask) Run(task *Task) (types.AnyType, types.BaseMethodFault) {
+	if c.VirtualMachine.Runtime.PowerState == c.state {
```

Comment:
So currently this can transition from a powered off state to suspended? Is that valid?